geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/TiledRasterLayer.scala

```diff
@@ -93,6 +93,45 @@ abstract class TiledRasterLayer[K: SpatialComponent: JsonFormat: ClassTag: Bound
     mask(geometries)
   }
 
+  def mask(
+    wkbRDD: RDD[Array[Byte]],
+    partitionStrategy: PartitionStrategy,
+    options: Rasterizer.Options
+  ): TiledRasterLayer[K] = {
+    val geometryRDD: RDD[Geometry] = wkbRDD.map { WKB.read }
+    val clipped: RDD[(SpatialKey, Geometry)] = geometryRDD.clipToGrid(rdd.metadata.layout)
+
+    val addValue = (arr: ArrayBuffer[Geometry], g: Geometry) => g +=: arr
+    val combineArrs = (arr1: ArrayBuffer[Geometry], arr2: ArrayBuffer[Geometry]) =>
+      if (arr1.size > arr2.size) arr2 ++=: arr1 else arr1 ++=: arr2
+
+    val aggregatedRDD: RDD[(SpatialKey, Seq[Geometry])] =
+      rdd.partitioner match {
+        case Some(p) =>
+          clipped.aggregateByKey(ArrayBuffer[Geometry](), p)(addValue, combineArrs).mapValues { _.toSeq }
```

Comment:
This seems like its trying to avoid calling `.groupByKey`, which would require less setup. Why is that?

Now that I look at it, it also seems like the `.groupByKey` follewed by join can be replaced by a single [`cogroup`](https://spark.apache.org/docs/2.1.1/api/scala/index.html#org.apache.spark.rdd.PairRDDFunctions@cogroup[W1,W2](other1:org.apache.spark.rdd.RDD[(K,W1)],other2:org.apache.spark.rdd.RDD[(K,W2)],partitioner:org.apache.spark.Partitioner):org.apache.spark.rdd.RDD[(K,(Iterable[V],Iterable[W1],Iterable[W2]))])

The left side of the tuple, should only contain a single tile, the right side will contain all the geometries for that tile.

Reply:
Also the shuffle here should either use the partitioning strategy, which it can because after clipping its indexed by key or it should use the partitioner of the tile RDD if no strategy is specified.

Reply:
My reasoning for doing an `aggregateByKey` was to avoid some of the cost of the shuffle brought on by a `groupByKey`. The reason for why I'd have it use the partitioner of the

Reply:
Gotcha.
Both [`groupByKey`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala#L498-L508) and [`aggregateByKey`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala#L160-L174) end up calling [`combineByKeyWithClassTag`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala#L71-L86).

The only difference is that `aggregateByKey` gives you a chance to provide an aggregating function that will be used in map-side combine where is `groupByKey` always concatenates all records for a key into a list.
 You would gain a benefit using `aggregateByKey` if the result of the combine would in some way be smaller than the individual parts (for instance merging two tiles into one). But thats not the case here, you're must be maintaining a list.